File: android/app/src/main/java/fyi/ikigai/memories/RNSharedWidget.java
package fyi.ikigai.memories;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Log;
import android.widget.ImageView;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

public class RNSharedWidget  extends ReactContextBaseJavaModule {
    ReactApplicationContext context;

    public RNSharedWidget(ReactApplicationContext reactContext) {
        super(reactContext);
        context = reactContext;
    }
    @NonNull
    @Override
    public String getName() {
        return "RNSharedWidget";
    }

    @ReactMethod
    public void setData(String key, String data, Callback callback) {
        SharedPreferences.Editor editor = context.getSharedPreferences("DATA", Context.MODE_PRIVATE).edit();
        editor.putString(key, data);
        editor.commit();

        Intent intent = new Intent(getCurrentActivity().getApplicationContext(), MemoriesWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        int[] ids = AppWidgetManager.getInstance(getCurrentActivity().getApplicationContext()).getAppWidgetIds(new ComponentName(getCurrentActivity().getApplicationContext(), MemoriesWidget.class));
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
        getCurrentActivity().getApplicationContext().sendBroadcast(intent);
        ((ImageView) getCurrentActivity().findViewById(R.id.image)).setScaleType(ImageView.ScaleType.FIT_CENTER);
        callback.invoke("done");
    }
}

File: android/app/src/main/java/fyi/ikigai/memories/RNSharedWidget.kt
package fyi.ikigai.memories

import NewMemoriesWidget
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetManager
import androidx.glance.appwidget.state.updateAppWidgetState
import androidx.glance.state.PreferencesGlanceStateDefinition
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import org.json.JSONObject

class RNSharedWidget(var context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "RNSharedWidget"
    }

    private val coroutineScope = MainScope()

    private val glanceAppWidget: GlanceAppWidget = NewMemoriesWidget()

    @ReactMethod
    fun setData(key: String?, data: String?, callback: Callback) {
        val json = data?.let { JSONObject(it) }

        coroutineScope.launch {
            val glanceId = GlanceAppWidgetManager(context).getGlanceIds(NewMemoriesWidget::class.java)
                .firstOrNull()

            glanceId?.let {
                updateAppWidgetState(context, PreferencesGlanceStateDefinition, it) { pref ->
                    pref.toMutablePreferences().apply {
                        if (json != null) {
                            this[currentName] = json.getJSONObject("value").getString("name")
                            this[currentDate] = json.getJSONObject("value").getString("start_datetime")
                            this[currentDistance] = json.getJSONObject("value").getString("distance_in_meters")
                            this[currentHeight] = json.getJSONObject("value").getString("total_elevation_gain_in_meters")
                            this[currentImageUrl] = json.getJSONObject("value").getString("picture_url")
                        }
                    }
                }
                glanceAppWidget.update(context, it)
            }
        }
    }

    companion object {
        val currentName = stringPreferencesKey("name")
        val currentDate = stringPreferencesKey("date")
        val currentDistance = stringPreferencesKey("distance")
        val currentHeight = stringPreferencesKey("height")
        val currentImageUrl = stringPreferencesKey("url")
    }
}
